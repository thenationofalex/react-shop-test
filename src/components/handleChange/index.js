export default function handleChange (e) {
  const target = e.target
  const value = target.value
  const name = target.id
  console.log(name, value)
  this.setState({
    [name]: value.toString()
  })
}
