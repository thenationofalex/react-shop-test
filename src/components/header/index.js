import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'

import { StyledHeader } from './styles'

class Header extends PureComponent {
  render () {
    return (
      <StyledHeader>
        <h1>{ this.props.title }</h1>
        { this.props.children }
      </StyledHeader>
    )
  }
}

Header.propTypes = {
  title: PropTypes.string,
  children: PropTypes.any
}

export default Header
