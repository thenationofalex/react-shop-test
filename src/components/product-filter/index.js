import React, { Fragment, PureComponent } from 'react'
import PropTypes from 'prop-types'
import flattenDeep from 'lodash/flattenDeep'
import uniq from 'lodash/uniq'

import { StyledSelect } from './styles'

let filterOptions = []

class ProductFilter extends PureComponent {
  render () {
    const { options, onChange } = this.props
    options.map(p => filterOptions.push(p.size))
    filterOptions = uniq(flattenDeep(filterOptions))
    return (
      <Fragment>
        <StyledSelect id='size' onChange={onChange}>
          <option value='reset'>Filter by size</option>
          {
            filterOptions.map(o => <option key={o} value={o}>{o}</option>)
          }
        </StyledSelect>
      </Fragment>
    )
  }
}

ProductFilter.propTypes = {
  options: PropTypes.array,
  onChange: PropTypes.func
}

export default ProductFilter
