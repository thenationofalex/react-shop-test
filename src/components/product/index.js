import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'

import {
  StyledProduct,
  StyledProductInfo,
  StyledLabelBlock,
  StyledSale,
  StyledExclusive
} from './styles'

class Product extends PureComponent {
  render () {
    const { isSale, isExclusive, price, productImage, productName } = this.props
    return (
      <StyledProduct>
        <img src={`/static/${productImage}`} alt={productName} />
        <StyledLabelBlock>
          {
            isSale && (<StyledSale>sale</StyledSale>)
          }
          {
            isExclusive && (<StyledExclusive>exclusive</StyledExclusive>)
          }
        </StyledLabelBlock>
        <StyledProductInfo>
          <h2>{productName}</h2>
          <p>{price}</p>
        </StyledProductInfo>
      </StyledProduct>
    )
  }
}

Product.propTypes = {
  isSale: PropTypes.bool,
  isExclusive: PropTypes.bool,
  price: PropTypes.string,
  productImage: PropTypes.string,
  productName: PropTypes.string
}

export default Product
