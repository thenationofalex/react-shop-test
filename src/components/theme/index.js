export const theme = {
  colors: {
   green: '#009900',
   red: '#ff0000',
   blue: '#def1f4'
  }
}