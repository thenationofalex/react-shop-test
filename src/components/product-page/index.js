import React, { Fragment, PureComponent } from 'react'
import Head from 'next/head'
import PropTypes from 'prop-types'

import { StyledPageWrapper, StyledProductWrapper } from './styles'
import Header from '../header'
import ProductFilter from '../product-filter'
import Product from '../product'
import handleChange from '../handleChange'

class ProductPage extends PureComponent {
  handleChange = handleChange.bind(this)
  state = { productList: this.props.products }

  componentDidUpdate (prevProps, prevState) {
    if (prevState.size !== this.state.size) {
      let filteredItems = []
      if (this.state.size === 'reset') {
        filteredItems = this.props.products
      } else {
        this.props.products.forEach(product => {
          if (product.size.indexOf(this.state.size) !== -1) {
            filteredItems.push(product)
          }
        })
      }
      this.setState({ productList: filteredItems })
    }
  }
  render () {
    const { title } = this.props
    const { productList } = this.state
    return (
      <StyledPageWrapper>
        <Head>
          <title>{title}</title>
        </Head>
        <Fragment>
          <Header title={title}>
            <ProductFilter options={productList} onChange={this.handleChange} />
          </Header>
          <StyledProductWrapper>
            {
              productList.map(p => <Product key={p.index} {...p} />)
            }
          </StyledProductWrapper>
        </Fragment>
      </StyledPageWrapper>
    )
  }
}

ProductPage.propTypes = {
  title: PropTypes.string.isRequired,
  products: PropTypes.arrayOf(PropTypes.shape({
    index: PropTypes.number,
    isSale: PropTypes.bool,
    isExclusive: PropTypes.bool,
    price: PropTypes.string,
    productImage: PropTypes.string,
    productName: PropTypes.string,
    size: PropTypes.array
  }))
}

export default ProductPage
