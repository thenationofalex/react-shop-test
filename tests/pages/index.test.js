import React from 'react'
import { shallow, mount } from 'enzyme'
import toJson from 'enzyme-to-json'
import { ThemeProvider } from 'emotion-theming'
import Index from '../../pages'
import products from '../../static/products.json'
import { theme } from '../../src/components/theme'

const props = {
  title: 'Test title',
  products
}

describe('Product Page', () => {
  test('Renders', () => {
    const wrapper = shallow(<Index {...props} />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })
})
