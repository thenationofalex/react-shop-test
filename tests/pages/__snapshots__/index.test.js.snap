// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Page Renders 1`] = `
<Fragment>
  <ProductPage
    products={
      Array [
        Object {
          "index": 0,
          "isExclusive": false,
          "isSale": true,
          "price": "$18.88",
          "productImage": "product-1.jpg",
          "productName": "Striped shirt",
          "size": Array [
            "XS",
            "S",
            "L",
            "XL",
          ],
        },
        Object {
          "index": 1,
          "isExclusive": false,
          "isSale": false,
          "price": "$25.44",
          "productImage": "product-2.jpg",
          "productName": "Denim shirt",
          "size": Array [
            "XS",
            "S",
          ],
        },
        Object {
          "index": 2,
          "isExclusive": true,
          "isSale": false,
          "price": "$12.93",
          "productImage": "product-3.jpg",
          "productName": "Plain cotton t-shirt",
          "size": Array [
            "S",
            "M",
          ],
        },
        Object {
          "index": 3,
          "isExclusive": false,
          "isSale": false,
          "price": "$26.32",
          "productImage": "product-4.jpg",
          "productName": "Plain 3/4 sleeve cotton t-shirt",
          "size": Array [
            "XL",
          ],
        },
        Object {
          "index": 4,
          "isExclusive": false,
          "isSale": false,
          "price": "$12.16",
          "productImage": "product-5.jpg",
          "productName": "White dress shirt",
          "size": Array [
            "M",
            "L",
          ],
        },
        Object {
          "index": 5,
          "isExclusive": false,
          "isSale": false,
          "price": "$30.27",
          "productImage": "product-6.jpg",
          "productName": "Long Sleeve Skivvy Top",
          "size": Array [
            "XS",
            "S",
            "M",
          ],
        },
        Object {
          "index": 6,
          "isExclusive": false,
          "isSale": true,
          "price": "$24.26",
          "productImage": "product-7.jpg",
          "productName": "Puffer Vest with Hood",
          "size": Array [
            "M",
            "L",
            "XL",
          ],
        },
        Object {
          "index": 7,
          "isExclusive": true,
          "isSale": false,
          "price": "$17.73",
          "productImage": "product-8.jpg",
          "productName": "Funnel Neck Swing Top",
          "size": Array [
            "XS",
            "S",
            "XL",
          ],
        },
      ]
    }
    title="Women's Tops"
  />
</Fragment>
`;
